### PR TITLE
Add support for reporting DNS resolution times

### PR DESCRIPTION
**Pull request by mem**
The time it takes to resolve hostnames is missing from the reported metrcs. Add support for recording that information thru the httptrace DNSStart and DNSDone hooks. This requires changing the signature of the netext.MultiResolver type because nettrace depends on having the hooks passed in a structure in the request context (they are wired up by httptrace).

### CODE REVIEW

**Comment by mem** [comment]
Tests are broken because I want some feedback before I go figuring out how to fix those.

File: lib/netext/httpext/tracer.go [hunk 1]
```diff
@@ -172,6 +186,14 @@ func (t *Tracer) Trace() *httptrace.ClientTrace {
 	}
 }
 
+func (t *Tracer) Start() {
+	t.start = now()
+}
+
+func (t *Tracer) End() {
```
Comment (mem) [inline lib/netext/httpext/tracer.go:193]: please ignore this, it's not part of this PR

File: lib/netext/httpext/tracer.go [hunk 2]
```diff
@@ -172,6 +186,14 @@ func (t *Tracer) Trace() *httptrace.ClientTrace {
 	}
 }
 
+func (t *Tracer) Start() {
+	t.start = now()
```
Comment (mem) [inline lib/netext/httpext/tracer.go:190]: please ignore this, this is not part of this PR

**Comment by na--** [comment]
Thanks, @mem! I removed myself as a reviewer, since I won't have any free time in the upcoming week and a half, but I can share some historical context here. 

First, we have an open issue about this, https://github.com/grafana/k6/issues/1011. And, as you can see in it, k6 used to measure DNS lookup times a long time ago, but then it was removed in https://github.com/grafana/k6/commit/21be5917036e59f09187aa14fbd67fc5da5f3159. The removal happened before my time, but from what I can see in the commit and the issue (https://github.com/grafana/k6/issues/41), it was done because we added some pretty heavy DNS caching. 

Since then, we have overhauled that caching pretty significantly, making it more configurable and usable. @imiric did the vast majority of the work (https://github.com/grafana/k6/pull/1612) here, that is why I added him to the PR. However, we didn't bring back the dns resolution metric at the time, so https://github.com/grafana/k6/issues/1011 remained open. 

There seemingly wasn't a ton of demand for that feature, and there is some complexity in doing this correctly, so we probably decided to cut the scope at the time. Can you explain your use specific case and why you need this in the issue?

Finally, a few words regarding the proposed implementation. `httptrace` may look like an easy place to measure this, but it will only work for HTTP requests and maybe WebSockets. It won't work for gRPC, Redis, or any other protocols we will want to introduce in the future, unless we figure out how to add it for every one of them too. Also, `httptrace` probably won't "understand" when k6 returns a cached result according to its [`--dns` resolution options](https://k6.io/docs/using-k6/k6-options/reference/#dns) and when it returns a fresh one, i.e. you are probably going to see a lot of `0` values measured due to instantaneous results.

So, it seems to me that the best place to add this metric would be the [`lib/netext/resolver.go`](https://github.com/grafana/k6/blob/master/lib/netext/resolver.go#L69-L70), though I am not sure what complications that would have, @imiric would know better than me. The one I can think of is that, while we know if k6 has cached the DNS result there, we don't actually know whether Go or the OS have cached it beyond us, so we will probably still get a lot of near-0 values for this new metric :disappointed: To fully prevent that and only measure non-cached lookups, we probably need https://github.com/grafana/k6/issues/1751, which is a much bigger project. Though I've noticed that the Go team are constantly making improvements to the native Go DNS resolver (e.g. see [Go 1.19](https://tip.golang.org/doc/go1.19#net) and [1.20](https://tip.golang.org/doc/go1.20#net) changelogs), so maybe something has changed and we wouldn't need to implement it from scratch :thinking:

**Comment by mem** [comment]
Hi @na-- 

thanks for taking the time to look at this. And thanks for digging up the the PRs and issues.

> Finally, a few words regarding the proposed implementation. `httptrace` may look like an easy place to measure this, but it will only work for HTTP requests and maybe WebSockets. It won't work for gRPC, Redis, or any other protocols we will want to introduce in the future, unless we figure out how to add it for every one of them too. Also, `httptrace` probably won't "understand" when k6 returns a cached result according to its [`--dns` resolution options](https://k6.io/docs/using-k6/k6-options/reference/#dns) and when it returns a fresh one, i.e. you are probably going to see a lot of `0` values measured due to instantaneous results.

Yeah, I probably worked at this backwards: I added the missing hook to the http tracer, and it didn't work (it was never called), and I started to look at why it didn't work. Since I'm familiar with that implementation, it was relatively straightforward to figure out what was going on: name resolution is started by the HTTP client, and the implementation is such that the context that the resolver ends up using doesn't have the hooks in place. I did think about modifying the resolver so that it performs the measurement itself, but I couldn't quite spot how to hook it to the correct location to actually add the observations. I probably need to spend more time with this code.

Regarding the aggressive caching that is in place... I'm probably missing something here, but it seems to me that the caching layer is ignoring the TTL in the DNS responses, and in particular it seems to be ignoring the case where the DNS server is specifying that the answer should not be cached... because it's not talking to a DNS server in the first place (if it doesn't have a cached response it's falling back to the system's name resolution mechanism)?

So you are right, there will be cases where the timings for DNS resolution are very low. As long as the caching layer does as asked, my opinion is that that's fine. That possibly mimicks the normal operation of a system correctly (some responses are locally cached, some responses are cached at the system level, etc). For example, my work computer has a caching forwarder in place that is in turn talking to another local caching forwarder.

(As a side note, the reason why I ended up here is because I'm interested in reporting end-to-end timings, i.e. the time `http.Do(req)` takes to complete, and I noticed a large discrepancy between that duration and the duration reported by `http_req_duration` -- I think that's the name -- and I was trying to figure out where that is coming from)

> So, it seems to me that the best place to add this metric would be the [`lib/netext/resolver.go`](https://github.com/grafana/k6/blob/master/lib/netext/resolver.go#L69-L70), though I am not sure what complications that would have, @imiric would know better than me.

This part makes sense. I would just need to figure out the mechanics of getting the observations to the correct place.

> The one I can think of is that, while we know if k6 has cached the DNS result there, we don't actually know whether Go or the OS have cached it beyond us, so we will probably still get a lot of near-0 values for this new metric disappointed To fully prevent that and only measure non-cached lookups, we probably need #1751, which is a much bigger project. Though I've noticed that the Go team are constantly making improvements to the native Go DNS resolver (e.g. see [Go 1.19](https://tip.golang.org/doc/go1.19#net) and [1.20](https://tip.golang.org/doc/go1.20#net) changelogs), so maybe something has changed and we wouldn't need to implement it from scratch thinking

This part I'm not fully following. I'm probably missing the intention.

If you want to avoid caches, you have to go ask the DNS server and honor the TTL in the response. I'm pretty sure that would break in many situations (e.g. netadmins not wanting people to talk directly to DNS servers and enforcing the use of local _caching_ forwarders; that's solved by specifying the DNS server to use).

So the part I'm confused about is where Microsoft Windows comes into play...

**Comment by mem** [comment]
Thank you for the discussion, this requires more planning to get the issue fixed.
